aquadoggo/src/replication/service.rs

```diff
@@ -0,0 +1,398 @@
+                    ));
+                }
+
+                if result.is_done {
+                    self.on_replication_finished(peer_id, connection_id, session_id)
+                        .await;
+                }
+            }
+            Err(err) => {
+                self.on_replication_error(peer_id, connection_id, session_id, err)
+                    .await;
+            }
+        }
+    }
+
+    async fn on_replication_finished(
+        &mut self,
+        peer_id: PeerId,
+        _connection_id: ConnectionId,
+        _session_id: SessionId,
+    ) {
+        info!("Finished replication with peer {}", peer_id);
+        match self.connections.get_mut(&peer_id) {
+            Some(peer) => {
+                peer.status.successful_count += 1;
+            }
+            None => {
+                panic!("Tried to access unknown peer");
+            }
+        }
+    }
+
+    async fn on_replication_error(
+        &mut self,
+        peer_id: PeerId,
+        connection_id: ConnectionId,
+        session_id: SessionId,
+        error: ReplicationError,
+    ) {
+        warn!("Replication with peer {} failed: {}", peer_id, error);
+
+        match self.connections.get_mut(&peer_id) {
+            Some(peer) => {
+                peer.status.failed_count += 1;
+            }
+            None => {
+                panic!("Tried to access unknown peer");
+            }
+        }
+
+        match error {
+            ReplicationError::StrategyFailed(_) | ReplicationError::Validation(_) => {
+                let peer_connection_id = PeerConnectionId(peer_id, Some(connection_id));
+                self.sync_manager
+                    .remove_session(&peer_connection_id, &session_id);
+            }
+            _ => (), // Don't try and close the session on other errors as it should not have been initiated
+        }
+    }
+
+    async fn handle_service_message(&mut self, message: ServiceMessage) {
+        match message {
+            ServiceMessage::ConnectionEstablished(peer_id, connection_id) => {
+                self.on_connection_established(peer_id, connection_id).await;
+            }
+            ServiceMessage::ConnectionClosed(peer_id, connection_id) => {
+                self.on_connection_closed(peer_id, connection_id).await;
+            }
+            ServiceMessage::ReceivedReplicationMessage(peer_id, connection_id, message) => {
+                self.on_replication_message(peer_id, message, connection_id)
+                    .await;
+            }
+            ServiceMessage::InitiateReplication => {
+                self.update_sessions().await;
+            }
+            _ => (), // Ignore all other messages
+        }
+    }
+
+    fn send_service_message(&self, message: ServiceMessage) {
+        if self.tx.send(message).is_err() {
+            // Silently fail here as we don't care if the message was received at this
+            // point
+        }
+    }
+
+    async fn update_sessions(&mut self) {
```

Comment:
Need to revisit this logic to check it's still doing what we expect (start a session for all connections we know about)

Reply:
Probably the maximum needs to be set to 1 right now since we're always using the same target set for each peer? So any other second replication session would be considered invalid (since it already exists)?

Think we're never seeing that problem since our replication is too fast with that little data.

Reply:
Also, fixed that filter logic issue:

```rust
let active_sessions: Vec<&Session> = sessions
    .iter()
    .filter(|session| !session.is_done())
    .collect();
```

.. where it only returned finished sessions before.

Reply:
> Probably what we need here is to check if that target set is already currently being replicated with that peer. We want max. one session per target set per peer, independent of established locally or remotely.

I've implemented that now